### ISSUE TITLE
Some confusion about the fast grads calculation when converting to Pytorch.

### ISSUE THREAD

**Issue by ShunLu91** [idiap/importance-sampling#36]
Hello, 

Great thanks to you for your great efforts. After reading your paper and code, I found that it's in fact a nice and solid work and I really enjoy it. 

To utilize this method in my model training, I try to implement your method using the Pytorch framework. I notice that you use the following code to calculate the gradient norm in a fast mode:
```
if self.fast:
    grads = K.sqrt(sum([
        self._sum_per_sample(K.square(g))
        for g in K.gradients(losses, self.parameter_list)
    ]))
```
As far as I am concerned, this line of the code [self._sum_per_sample(K.square(g)) for g in K.gradients(losses, self.parameter_list)] has computed the gradients square and summed them per sample. I am confused about why not directly use K.sqrt() function to get the gradient norm of each sample but introduce another sum() function behind the K.sqrt()? 

Besides, I have checked the results of sum([self._sum_per_sample(K.square(g)) for g in K.gradients(losses, self.parameter_list)]) and [self._sum_per_sample(K.square(g)) for g in K.gradients(losses, self.parameter_list)], and found that they were equal, which is really amazing. And if I remove the sum() function behind the K.sqrt(), it will raise the data type error. Therefore, does this sum() function only convert the data type and not perform summation?

Expect your reply and I will share my Pytorch implementation once they are ready.

Best, 
Shun Lu

**Comment by 3DJakob** [comment]
@ShunLu91 Did you eventually release your pytorch code? I would be very interested in looking at it.

**Comment by ShunLu91** [comment]
@3DJakob Sorry for the late reply and we have released our code just now. Our method proposed to explicitly minimize the gradient variance of the supernet training by jointly optimizing the sampling distributions of PAth and DAta, namely [PA&DA](https://github.com/ShunLu91/PA-DA). 
For the essential part of calculating the grad of each data, we finally take the reference from [RHO-Loss](https://github.com/OATML/RHO-Loss) and use it [here](https://github.com/ShunLu91/PA-DA/blob/887b6ed0ff3fc4e02f55d2f8dcec567ca33ff32e/nasbench201/train_panda_201.py#L134). 
Welcome to try our code and discuss with us.

**Comment by 3DJakob** [comment]
@ShunLu91 thank you for your response I will take a look at it 👍🏻
